src/ml/trainer/JasmineGraphTrainingSchedular.cpp

```diff
@@ -177,3 +178,215 @@ JasmineGraphTrainingSchedular::packPartitionsToMemory(vector<pair<int, int>> par
+            partitionID = stoi(id_partition);
+            vertexcount = stoi(vertexCount) + stoi(centralvertexCount);
+            edgecount = stoi(edgeCount) + stoi(centralEdgeCount);
+
+            vector<int> partitionCounts; //Vector for partition node count, edge count and feature count
+            partitionCounts.push_back(partitionID);
+            partitionCounts.push_back(vertexcount);
+            partitionCounts.push_back(edgecount);
+            partitionCounts.push_back(featurecount);
+
+            partitionDetails.push_back(partitionCounts);
+
+            //Store assigned worker per partition
+            workerid = stoi(rowData.at(5).second);
+            partitionWorkerMap[partitionID] = workerid;
+
+        }
+
+        long availableMemory = getAvailableMemory(j->second); //Host memory
+
+        //Get memory estimation list for each partition of host
+        vector<pair<int, int>> partitionMemoryList = estimateMemoryDistOpt(partitionDetails, availableMemory);
+
+        //Get schedule for host
+        map<int, map<int, int>> scheduledPartitionSets = bestFitPartition(partitionMemoryList, partitionWorkerMap, availableMemory);
+        scheduleForEachHost.insert(make_pair(j->second, scheduledPartitionSets));
+    }
+    return scheduleForEachHost;
+}
+
+/** Estimate memory of graph partitions using the partition details
+ *
+ * @param partitionDetails a vector of vectors containing partition ID, node count, edge count, feature count per partition
+ * @param availableMemory total memory of host
+ */
+vector<pair<int, int>>
+JasmineGraphTrainingSchedular::estimateMemoryDistOpt(vector<vector<int>> partitionDetails, long availableMemory) {
+
+
+    vector<pair<int, int>> partitionMemoryList; //Vector of estimated sizes of partitions
+
+    //Iterate through partitions and estimate sizes
+    for (vector<vector<int>>::iterator i = partitionDetails.begin(); i != partitionDetails.end(); i++) {
+        int partitionID = (*i)[0];
+        int nodeCount = (*i)[1];
+        int edgeCount = (*i)[2];
+        int featureCount = (*i)[3];
+
+        int graphSize = 8*nodeCount*edgeCount*featureCount / (1024 * 1024); //Size in hard disk
+
+        int sizeInMem = 36 * graphSize + 2; //Size when loaded into memory
+
+        if (sizeInMem > availableMemory) {
+            sizeInMem = availableMemory;
+        }
+
+        partitionMemoryList.push_back(make_pair(partitionID, sizeInMem));
+    }
+
+    return partitionMemoryList;
+}
+
+/** Create schedule for host by best fitting in sorted order (descending)
+ *
+ * @param partitionMemoryList Vector of pairs containing partition ID and estimated size in memory
+ * @param partitionWorkerMap Map of partition id and the worker assigned to partition
+ * @param capacity total memory of host
+ */
+std::map<int, map<int, int>>
+JasmineGraphTrainingSchedular::bestFitPartition(vector<pair<int, int>> partitionMemoryList, map<int, int> partitionWorkerMap, int capacity) {
```

Comment:
Can we have better descriptive function name than `bestFitPartition` ?

Reply:
Method name changed to `schedulePartitionsBestFit`

Reply:
Done